Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -209,11 +209,16 @@ public final actor SemanticIndexManager {
     for uri in changedFiles {
       indexStatus[uri] = nil
     }
-    // Clear the preparation status so that we re-prepare them. If the target hasn't been affected by the file changes,
-    // we rely on a null build during preparation to fast re-preparation.
-    // Ideally, we would have more fine-grained dependency management here and only mark those targets out-of-date that
-    // might be affected by the changed files.
-    preparationStatus.removeAll()
+    let targets = await changedFiles.asyncMap { await buildSystemManager.configuredTargets(for: $0) }.flatMap { $0 }
+    if let dependentTargets = await buildSystemManager.targets(dependingOn: targets) {
+      for dependentTarget in dependentTargets {
+        preparationStatus[dependentTarget] = nil
+      }
+    } else {
+      // We couldn't determine which targets depend on the modified targets. Be conservative and assume all of them do.
+      preparationStatus = [:]
```

Comment:
We could not remove `targets` in this case, not that it's really expected to happen anyway.

Reply:
The only case where we’d be able to not remove the target itself is if there is only a single target. Say you have a project two targets `A` and `B` and you change one file in `A` and one file in `B` then you don’t know if they depend on each other and you need to assume that they do. So you need to invalidate both.

I don’t think it’s worth optimizing for the single module case. Really, the build system should provide more accurate information here.

Reply:
Single target is the *main* case though 😅 I agree the buildsystem should, but asking for preparation on every edit is really pretty bad.

Reply:
In one of my follow-up PRs (not up yet), I’m re-implementing the entire up-to-date tracking of targets and that allows us to early-exit if we know that the target is up-to-date without any build system interaction.